docs/RFCS/20181209_lazy_txn_record_creation.md

```diff
@@ -0,0 +1,204 @@
+
+What this means in practice is that a non-replayed request from the transaction coordinator itself should never stumble into an already GCed transaction record without also finding itself beneath the `TxnSpanGCThreshold`. Requests could be replayed, but that is [protected against](#Replay-Protection-/-Idempotency) as well.
+
+Thing are more interesting for concurrent transactions, which have no guarantee about the state that they observe a transaction in. It is completely valid that a concurrent transaction observes a missing transaction record that was either `COMMITTED` or `ABORTED` and later GCed without having any indication of what happened to it. This is already the case, and we opt for rewriting the transaction record as `ABORTED` in that case. This isn't great, but it also won't cause any issues.
+
+The situation becomes a little worse now that `batcheval.PushTxn` will handle (often wait) for missing transaction records. In fact, a `PushTxn` request has no way of determining whether a transaction record has not yet been written or has already been finalized and GCed. This won't cause any correctness issues, but it means that naively a `PushTxn` may end up waiting for a transaction record that has already been GCed simply because it saw activity on one of the transaction's intents within the `TxnLivenessThreshold`.
+
+It is unclear how to avoid this issue entirely. This RFC's current thinking is that we should do away with the `TxnAutoGC` fast-path and enforce a `TxnLivenessThreshold` waiting period on eager transaction record GCs. If we made the eager GC of transaction records wait `TxnLivenessThreshold` then any `PushTxn` request in this period would find a non-GCed transaction record and any `PushTxn` request after this period would necessarily synthesize an expired transaction record. Thus, we would avoid the issue. This waiting period could be introduced in the `intentResolver`, and plays into some of the goals outlined in [#30780](https://github.com/cockroachdb/cockroach/issues/30780).
+
+### QueryTxn / txnwait.Queue
+
+The `txnwait.Queue` queries both the pusher and the pushee transaction using `QueryTxn` requests while waiting for changes to either. While doing so, the `txnwait.Queue` has access to `TxnMeta` objects for both of these transactions. Either the `QueryTxn` request or the `txnwait.Queue` itself will need to be made aware that transaction records are now creates lazily. 
+
+Currently the `txnwait.Queue` works properly even if the pusher does not have a transaction record but does not behave properly when the pushee does not have a transaction record. There are a few alternatives to fixing this:
+- `QueryTxn` could use its provided `TxnMeta` object to synthesize a transaction record when one does not exist. This is nice from the perspective of hiding complication behind an abstraction, but might undermine the purpose of `QueryTxn` in certain situations.
```

Comment:
I prefer the approach of `QueryTxn` synthesizing the txn record. Since it already must return the list of dependent txns, it is implicitly agreeing that the queried txn exists. Fulfilling the contract of the API call by positing the state of the implicit txn and returning it maintains the abstraction and seems easier to reason about. Letting this bit of complexity leak into `txnwait.Queue` feels worse.